Autospidergen - First pass of documentation (#10097)
alltheplaces/alltheplaces: docs/DETECTION_AND_CODE_GENERATION.md

## Detection

### Inspection of site

> Alltheplaces provides rudimentary tooling to inspect a site for common behaviours.

Author: ```suggestion
All The Places provides rudimentary tooling to inspect a site for common behaviours.
```

> Alltheplaces provides rudimentary tooling to inspect a site for common behaviours.

Author: ```suggestion
All The Places provides rudimentary tooling to inspect a site for common behaviours.
```

* [Sitemap inspection](SITEMAP.md) - `pipenv run scrapy sitemap http://example.com/` to detect potential individual store URLs.

> * [Structured Data inspection](STRUCTURED_DATA.md) - `pipenv run scrapy sitemap http://example.com/` to detect potential individual store URLs.

Author: ```suggestion
* [Structured Data inspection](STRUCTURED_DATA.md) - `pipenv run scrapy sd http://example.com/` to detect potential individual store URLs.
```

* Links - `pipenv run scrapy links http://example.com/` to look for links with human language labels, such as "Find our stores".

When checking a large number of URLs, these tools can be chained together to highlight potential candidates for spidering.

### Automatic detection from storefinder page

> 

Author: ```suggestion

There exist numerous third-party software-as-a-service APIs or self-installed web application software which are used by brands for providing a store finder or stock finder page on brand websites. All The Places implements a "storefinder" as a generic pattern, allowing a spider for each brand to inherit this generic pattern without having to individually duplicate complex crawler code.

```

Alltheplaces has a number of key storefinders which gather together common functionality.

Going one step further, many of these store finders build in the capabilities to automatically detect the
presence of a common storefinder from either:

- Request patterns made to API endpoints or

> - Responses that contain xpath or JS Objects indicating the presence of a store locator.

Reviewer: Suggest "Responses in HTML format (DOM elements and JavaScript objects), including any part of a response which is loaded from external JavaScript files or from within iframes)"

> - Responses that contain xpath or JS Objects indicating the presence of a store locator.

Author: ```suggestion
- Responses where the HTML matches xpath or creates JS Objects indicating the presence of a store locator.
```

To automatically attempt to detect a storefinder, pass in a start URL - either the top level domain, or
the specific store page.
Hint: You may wish to use the `pipenv run scrapy links http://example.com/` command to automatically look for

> probably storefinder pages if you are reviewing a number of top level domains.

Reviewer: probably -> probable

Example:
```
ubuntu@codespaces-dc76c9:/workspaces/alltheplaces$ pipenv run scrapy sf https://www.saveonfoods.com/
from locations.storefinders.storefrontgateway import StorefrontgatewaySpider

class SaveOnFoodsCASpider(StorefrontgatewaySpider):
    name = "saveonfoods_ca"
    item_attributes = {
        "brand_wikidata": "Q7427974",
        "brand": "Save-On-Foods",
    }
    start_urls = [
        "https://storefrontgateway.saveonfoods.com/api/",
    ]
```

> This tool will attempt to generate a useful class name, spider key name, and

Reviewer: Suggest stating how the automatic class name, spider key name and brand name/Wikidata ID are generated from the URL. A link to the NSI project can be provided as the source of data used for this mapping--e.g. TLD for domain being scanned is checked against NSI database for a match, then NSI data is pulled for that TLD.

> This tool will attempt to generate a useful class name, spider key name, and

Author: ```suggestion
This tool will attempt to generate a useful class name, spider key name, by resolving the URL to many wikidata or name suggestion index entries it can. It will also extract
```

all other attributes to get you started; however you should still run and adjust the generated code
before opening a pull request.

> If known, you can specify extra arguments such as `--brand-wikidata`.

Reviewer: Perhaps include an example of a full command line, and also list all supported extra arguments that the `sf` command expects.

For full usage, run `pipenv run scrapy sf`

### Manual Generation

Where a storefinder isn't detected but you still wish to get a jump start on writing a spider to a common pattern,

> use the `pipenv run scrapy genspider` command to generate a common template.

Reviewer: How does someone specify a particular template to generate a spider for? Can an example be provided for generating a SitemapSpider + StructuredDataSpider and a StockistSpider (or whatever storefinder)?

Author: ```suggestion
use the `pipenv run scrapy genspider` command to generate a common template.

For example, `pipenv run scrapy genspider my_spider_name http://example.com/robots.txt -t sitemap` generates a sitemap spider for example.com with the name my_spider_name.
```

## Adding autodetection to a storefinder

> If you have identified a reliable pattern, to add auto detection use one or more of:

Reviewer: Suggest expanding to: "to add autodetection to one of All The Places' generic storefinders"

Also suggest noting that the storefinder class needs to subclass `AutomaticSpiderGenerator` and additionally specify the `detection_rules` attribute which is a list of one or more detection rules (`DetectionRequestRule` and `DetectionResponseRule`).

- DetectionRequestRule
- DetectionResponseRule

Example:
```
from locations.automatic_spider_generator import AutomaticSpiderGenerator, DetectionRequestRule, DetectionResponseRule

class MyNewStoreLocatorSpider(Spider, AutomaticSpiderGenerator):

>     detection_rules = [

Reviewer: These detection rules will need a bit of documentation to explain how each one works.

By introduction:
- Playwright with Firefox (Chromium not supported) is required and is used to load the requested URL in a Firefox headless session.
- The Playwright Firefox session is kept alive for 30 seconds (or whatever we set) to observe any dynamic requests made by the page after the DOM has finished loading.
- The Playwright Firefox session is configured to automatically deny any requests for browser permissions such as requests for geolocation. This is necessary to ensure that dynamic loading of content is not blocked by a browser permission request that is never responded to by a user.
- Automatic redirects are followed in the Playwright Firefox session.
- All detection rules are executed against all requests made in the Playwright Firefox session and all iframes which are loaded in the Playwright Firefox session.
- An unsuccessful detection can be caused by geoblocking, anti-bot detection and CAPTCHA challenges and blocking modal dialogs that brand websites may present requiring user interaction before the page loads. If a storefinder is not automatically detected, there could still be a storefinder in use that can be observed in a normal interactive web browser session using built-in browser debugging tools.

Detail for each type of rule:
- URL parameter extraction with PCRE named capture groups
- HTTP header parameter extraction with PCRE named capture groups
- JavaScript lambda execution to return objects (after dynamic loading is complete)
- HTML XPATH query to extract strings from DOM objects (after dynamic loading is complete)
- JQ language query to extract objects from a JSON object (such as the body of a HTTP POST request)
- ...etc

Then description of attribute typing with use of the `__list` suffix (etc) and matching but not extracting with use of the `__` prefix.
